test/python/transpiler/test_linear_functions_passes.py

```diff
@@ -0,0 +1,160 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test transpiler passes that deal with linear functions."""
+
+import unittest
+
+from qiskit.circuit import QuantumCircuit
+from qiskit.transpiler.passes.optimization import CollectLinearFunctions
+from qiskit.transpiler.passes.synthesis import (
+    LinearFunctionsSynthesis,
+    LinearFunctionsToPermutations,
+)
+from qiskit.test import QiskitTestCase
+from qiskit.circuit.library.generalized_gates import LinearFunction
+from qiskit.transpiler import PassManager
+from qiskit.quantum_info import Operator
+
+
+class TestLinearFunctionsPasses(QiskitTestCase):
+    """Tests to verify correctness of the transpiler passes that deal with
+    linear functions:
+    the pass that extracts blocks of CX and SWAP gates and replaces these blocks by LinearFunctions,
+    the pass that synthesizes LinearFunctions into CX and SWAP gates,
+    and the pass that promotes LinearFunctions to Permutations whenever possible.
+    """
+
+    def test_single_linear_block(self):
+        """Test that when all gates in a circuit are either CX or SWAP,
+        we end up with a single LinearFunction."""
+
+        # original circuit
+        circuit = QuantumCircuit(4)
+        circuit.cx(0, 1)
+        circuit.cx(0, 2)
+        circuit.cx(0, 3)
+        circuit.swap(2, 3)
+        circuit.cx(0, 1)
+        circuit.cx(0, 3)
+
+        # new circuit with linear functions extracted using transpiler pass
+        optimized_circuit = PassManager(CollectLinearFunctions()).run(circuit)
+
+        # check that this circuit consists of a single LinearFunction
+        self.assertTrue("linear_function" in optimized_circuit.count_ops().keys())
+        self.assertTrue(len(optimized_circuit.data) == 1)
+        inst1, _, _ = optimized_circuit.data[0]
+        self.assertTrue(isinstance(inst1, LinearFunction))
+
+        # construct a circuit with linear function directly, without the transpiler pass
+        expected_circuit = QuantumCircuit(4)
+        expected_circuit.append(LinearFunction(circuit), [0, 1, 2, 3])
+
+        # check that we have an equivalent circuit
+        self.assertEqual(Operator(optimized_circuit), Operator(expected_circuit))
+
+        # now a circuit with linear functions synthesized
+        synthesized_circuit = PassManager(LinearFunctionsSynthesis()).run(optimized_circuit)
+
+        # check that there are no LinearFunctions present in synthesized_circuit
+        self.assertFalse("linear_function" in synthesized_circuit.count_ops().keys())
+
+        # check that we have an equivalent circuit
+        self.assertEqual(Operator(optimized_circuit), Operator(synthesized_circuit))
+
+    def test_two_linear_blocks(self):
```

Comment:
It would be good to put in some more tests of potential edge cases in block collection - things like if the first or last blocks are non-linear, if there's a run of non-linear functions in the middle, and if a block will simplify down to the identity matrix.  I think it should work with them already, but it'll give the next person to change the code more confidence that they haven't broken something.

(Come to think of it, weird ways of spelling the identity matrix would be good tests for the synthesis routine as well.)